### PR TITLE
Improved error message for PasswordEncoder

### PR DESCRIPTION
**Pull request by bottlerocketjonny**
I fear I might have been too slow to submit this! However, this closes gh-14880

Happy to make any changes.

### CODE REVIEW

**Comment by jzheaux** [comment]
Hi, @bottlerocketjonny! Yes, we already received another PR for the same and I think it is a nice improvement. That said, if you'd like, please rebase once that is merged and push any other changes you'd like to see to this PR, and we can take a look together.

**Comment by bottlerocketjonny** [comment]
> Hi, @bottlerocketjonny! Yes, we already received another PR for the same and I think it is a nice improvement. That said, if you'd like, please rebase once that is merged and push any other changes you'd like to see to this PR, and we can take a look together.

No worries! Yeah sure thing I shall do that 👍🏻

Just for future reference, if I wanted to work on an issue should I wait to be assigned or just go for it? I left a comment on this one but wasn't sure if that was the way to go. Cheers 😀

**Comment by jzheaux** [comment]
@bottlerocketjonny, it's certainly better when I'm on top of folks' comments in tickets so that there isn't confusion. :) Sorry that you and the other contributor ended up working on the same feature.

It's rare for two people to contribute near the same time, so we don't have a stated policy. Mostly, it's a best effort on my part to notice the comment, assign the issue to that person, and in the absence of that, attend to the first submitted PR.

In your case, I'm happy that there were good ideas in both PRs, so I'd still encourage you to rebase.

File: crypto/src/main/java/org/springframework/security/crypto/password/DelegatingPasswordEncoder.java [hunk 1]
```diff
@@ -245,12 +245,12 @@ private String extractId(String prefixEncodedPassword) {
 			return null;
 		}
 		int start = prefixEncodedPassword.indexOf(this.idPrefix);
-		if (start != 0) {
+		if (start == -1 && !prefixEncodedPassword.contains(this.idSuffix)) {
```
Comment (jzheaux) [inline crypto/src/main/java/org/springframework/security/crypto/password/DelegatingPasswordEncoder.java:248]: I'd like to be a bit more conservative with changing this behavior, given that it is called by more than one flow in this class. What if the `matches` implementation itself did the needed checking? Something like this:

```java
String id = extractId(prefixEncodedPassword);
if (has a value) {
    throw no mapping exception
}
start = indexof idPrefix
end = indexof idSuffix, start
if (start < 0 && end < 0) {
    throw no encoder prefix 
}
throw malformed encoder prefix
```

In exchange for a small amount of duplication, we don't change the expectations of other flows calling `extractId`.

File: crypto/src/main/java/org/springframework/security/crypto/password/DelegatingPasswordEncoder.java [hunk 2]
```diff
@@ -286,7 +286,16 @@ public String encode(CharSequence rawPassword) {
 		@Override
 		public boolean matches(CharSequence rawPassword, String prefixEncodedPassword) {
 			String id = extractId(prefixEncodedPassword);
-			throw new IllegalArgumentException("There is no PasswordEncoder mapped for the id \"" + id + "\"");
+			if (id == null) {
+				throw new IllegalArgumentException("Password encoding information cannot be null. " +
```
Comment (jzheaux) [inline crypto/src/main/java/org/springframework/security/crypto/password/DelegatingPasswordEncoder.java:290]: I think another bit of advice is needed here. Something like:

```
Given that there is no default password encoder configured, each password must have a password encoding prefix. Please either prefix this password with '{noop}' or set a default password encoder in `DelegatingPasswordEncoder`.
```

File: core/src/test/java/org/springframework/security/provisioning/InMemoryUserDetailsManagerTests.java [hunk 3]
```diff
@@ -18,15 +18,24 @@
 
 import java.util.Properties;
 
+import java.util.stream.Stream;
```
Comment (jzheaux) [inline core/src/test/java/org/springframework/security/provisioning/InMemoryUserDetailsManagerTests.java:21]: Please update the copyright here to `2002-2024`

File: crypto/src/main/java/org/springframework/security/crypto/password/DelegatingPasswordEncoder.java [hunk 4]
```diff
@@ -245,12 +245,12 @@ private String extractId(String prefixEncodedPassword) {
 			return null;
 		}
 		int start = prefixEncodedPassword.indexOf(this.idPrefix);
-		if (start != 0) {
```
Comment (jzheaux) [inline crypto/src/main/java/org/springframework/security/crypto/password/DelegatingPasswordEncoder.java:248]: Please update the copyright here to `2002-2024`

File: crypto/src/test/java/org/springframework/security/crypto/password/DelegatingPasswordEncoderTests.java [hunk 5]
```diff
@@ -193,31 +193,31 @@ public void matchesWhenNoopThenDelegatesToNoop() {
 	public void matchesWhenUnMappedThenIllegalArgumentException() {
 		assertThatIllegalArgumentException()
 			.isThrownBy(() -> this.passwordEncoder.matches(this.rawPassword, "{unmapped}" + this.rawPassword))
-			.withMessage("There is no PasswordEncoder mapped for the id \"unmapped\"");
```
Comment (jzheaux) [inline crypto/src/test/java/org/springframework/security/crypto/password/DelegatingPasswordEncoderTests.java:196]: Please update the copyright here to `2002-2024`

**Review by jzheaux — CHANGES_REQUESTED** [review]
Thanks for the updates, @bottlerocketjonny! I've left some feedback inline.

Also, will you please update your commit message to include the ticket https://github.com/spring-projects/spring-security/issues/14880 like so:

```
Your Title

Issue gh-14880
```

That will link this work to that ticket in GitHub, making it easier to track in the future.

**Comment by bottlerocketjonny** [comment]
@jzheaux Oh gosh, so sorry I completely missed that you'd replied here. I'll get this sorted ASAP so we can close this issue.

**Comment by bottlerocketjonny** [comment]
@jzheaux I've just pushed some new commits that I think are an improvement - let me know if anything else needs changing and apologies again for my tardiness! :)

**Review by jzheaux — APPROVED** [review]

**Comment by jzheaux** [comment]
Thanks, @bottlerocketjonny! This is now merged into `main`.

**Comment by bottlerocketjonny** [comment]
> Thanks, @bottlerocketjonny! This is now merged into `main`.

Excellent! Thanks for your patience.

**Pull request merged**